src/main/java/org/apache/commons/pool2/impl/EvictionTimer.java

```diff
@@ -129,4 +167,45 @@ public Thread newThread(final Runnable runnable) {
             return thread;
         }
     }
+
+    /**
+     * Task that removes references to abandoned tasks and shuts
+     * down the executor if there are no live tasks left.
+     */
+    private static class Reaper implements Runnable {
+        @Override
+        public void run() {
+            synchronized (EvictionTimer.class) {
+                for (Entry<WeakReference<Runnable>, WeakRunner> entry : taskMap.entrySet()) {
+                    if (entry.getKey().get() == null) {
+                        executor.remove(entry.getValue());
+                        taskMap.remove(entry.getKey());
+                    }
+                }

```

Comment:
Interesting observation.  I left it in after addressing your comment below because dead references (and empty executors) may hang around longer if we wait until they run to kill them.